### PR TITLE
Custom Dimensions

### PR DESCRIPTION
**Pull request by dotsbb**
Ref #70

### CODE REVIEW

**Comment by d4rken** [comment]
Maybe some javadoc to say that customvars are being deprecated at some point and some javadoc to say currently customdimensions require a plugin (insert plugin link).

**Comment by dotsbb** [comment]
@d4rken Sure, I'll update javadoc but I can't figure out how we can replace `CustomVariables`  with new plugin aproach 	though. Some users may want to sent *customvars* **and** *custom dimensions* in one request.

**Comment by d4rken** [comment]
AFAIK the new plugin approach is not there to replace the current approach. It's a switch that will be done with Piwik 3.0 and for now requires an extra plugin. So currently you can install the plugin and do both customvariables and customdimensions.

**Comment by d4rken** [comment]
The current custom variables approach does [json conversion](https://github.com/piwik/piwik-sdk-android/blob/dev/piwik-sdk/src/main/java/org/piwik/sdk/CustomVariables.java), so we can't really replace it anyways as the server wouldn't understand the syntax ```dimension$dimensionId=$dimensionValue``` without the plugin and even then, would it group the information in the same way? Probably not.

File: piwik-sdk/src/main/java/org/piwik/sdk/plugins/CustomDimensions.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ */
+public class CustomDimensions extends TrackMe {
+
+    /**
+     * This method sets a tracking API parameter dimension%dimensionId%=%dimensionValue%.
+     * Eg dimension1=foo or dimension2=bar.
+     * So the tracking API parameter starts with dimension followed by the set dimensionId.
+     * @return instance of CustomDimensions plugin
+     */
+    public synchronized CustomDimensions set(int dimensionId, String dimensionValue) {
```
Comment (d4rken) [inline piwik-sdk/src/main/java/org/piwik/sdk/plugins/CustomDimensions.java]: We should probably document the limitations here. See: https://github.com/piwik/plugin-CustomDimensions/issues/6

Actually we could enforce them.
* Check that the ```dimensionId```is valid. I guess it's 0-* or 1-*?
* ```dimensionValue``` is max 255 chars.

Sp if the paramaters are bad we could throw a warning and just not set them.

File: piwik-sdk/src/test/java/org/piwik/sdk/TrackerTest.java [hunk 2]
```diff
@@ -808,6 +809,19 @@ public void testPreviousVisit() throws Exception {
         assertEquals(1000l, Long.parseLong(queryParams.get(QueryParams.PREVIOUS_VISIT_TIMESTAMP)));
     }
 
+    @Test
+    public void testSetCustomDimensions() throws Exception {
+        CustomDimensions customDimensions = new CustomDimensions();
+        customDimensions.set(1, "foo");
+        customDimensions.set(2, "bar");
+        customDimensions.set(3, "empty").set(3, null);
```
Comment (d4rken) [inline piwik-sdk/src/test/java/org/piwik/sdk/TrackerTest.java]: Good check!
What is the expected behavior client/server side for an empty string ```""``` ?

File: piwik-sdk/src/main/java/org/piwik/sdk/plugins/CustomDimensions.java (hunk 1 above)
Comment (dotsbb) [inline piwik-sdk/src/main/java/org/piwik/sdk/plugins/CustomDimensions.java]: Good spotted.
`dimensionId` is 1+ https://github.com/piwik/plugin-CustomDimensions/blob/master/Dao/Configuration.php#L152

File: piwik-sdk/src/test/java/org/piwik/sdk/TrackerTest.java (hunk 2 above)
Comment (dotsbb) [inline piwik-sdk/src/test/java/org/piwik/sdk/TrackerTest.java]: An empty string behaves the same way as `null` value
https://github.com/piwik/plugin-CustomDimensions/blob/e330cb2c0db08d1a5b4e489d3fba7acc0e7cc43c/Tracker/CustomDimensionsRequestProcessor.php#L100

**Comment by d4rken** [comment]
Did you remove the javadoc changes to ```setScreenCustomVariable```? I think we could still encourage people to look at CustomDimensions. As deprecated might be too harsh, maybe something like: 

> This will be replaced by https://plugins.piwik.org/CustomDimensions at some point.

Otherwise, good job,  :+1:  for merge into dev.

btw what does the label CR stand for?

**Comment by dotsbb** [comment]
I have moved javadoc with `setScreenCustomVariable` straight to the master.
CR means that a Code Review is needed.

**Pull request merged**